src/dune_rules/foreign_sources.ml

```diff
@@ -124,23 +124,57 @@ let check_no_qualified (loc, include_subdirs) =
 
 let make (d : _ Dir_with_dune.t) ~(sources : Foreign.Sources.Unresolved.t)
     ~(lib_config : Lib_config.t) =
-  let libs, exes =
-    List.filter_partition_map d.data ~f:(fun stanza ->
+  let libs, foreign_libs, exes =
+    List.fold_left d.data ~init:([], [], [])
+      ~f:(fun ((libs, foreign_libs, exes) as acc) stanza ->
         match (stanza : Stanza.t) with
         | Library lib ->
           let all = eval_foreign_stubs d lib.buildable.foreign_stubs ~sources in
-          Left (Left (lib, all))
+          ((lib, all) :: libs, foreign_libs, exes)
         | Foreign_library library ->
           let all = eval_foreign_stubs d [ library.stubs ] ~sources in
-          Left (Right (library.archive_name, (library.archive_name_loc, all)))
-        | Executables exes ->
-          let all =
-            eval_foreign_stubs d exes.buildable.foreign_stubs ~sources
-          in
-          Right (exes, all)
-        | _ -> Skip)
+          ( libs
+          , (library.archive_name, (library.archive_name_loc, all))
+            :: foreign_libs
+          , exes )
+        | Executables exe
+        | Tests { exes = exe; _ } ->
+          let all = eval_foreign_stubs d exe.buildable.foreign_stubs ~sources in
+          (libs, foreign_libs, (exe, all) :: exes)
+        | _ -> acc)
+  in
+  let () =
+    let objects =
+      List.unordered_concat
```

Comment:
I'm not sure using the unordered variant is a good idea here: we'll end up with a hard to understand order of objects, which might end up in a link command line or error message or somewhere similarly bad (even if it's not the case now, it might happen after a future refactoring). All this just to avoid a couple of `List.rev`s? Just the mere existence of `List.unordered_concat` makes me nervous :)

Reply:
(Same concern about all the `List.rev_map`s below: this seems like a wrong thing to optimise in a build system.)

Reply:
P.S.: Actually, if this is indeed worth optimising (say, it shows up in profile), then we could introduce a separate data structure for managing (some) lists in Dune, which would provide stable order and fast `concat_map`.

Reply:
> hen we could introduce a separate data structure for managing (some) lists in Dune, which would provide stable order and fast concat_map.

Like a dlist? I suppose it's worth trying.

Amusingly, the error messages after this PR are actually improved in my view. The duplicate "thing" pointed out by the error is now the second rather than the first occurrence of the thing.

I'm still a bit annoyed by how we do all this extra work that is only useful in the tiny minority of cases when the error occurs.

Reply:
> P.S.: Actually, if this is indeed worth optimising (say, it shows up in profile), then we could introduce a separate data structure for managing (some) lists in Dune, which would provide stable order and fast concat_map.

In my profiling, a 1/3rd of all work in a zero build is just GC'ing. Last time I did some deforestation, I managed to shave off 100ms off a zero rebuild for some moderate workspace. Nothing to scoff at. Especially given that the potential to optimize more is huge given all that trash we currently produce.

I removed the optimizations for now.

Reply:
> Last time I did some deforestation, I managed to shave off 100ms off a zero rebuild for some moderate workspace. Nothing to scoff at.

I agree, it's definitely worth looking at more efficient data structures for various things (ordered lists, paths, etc.). I was just a bit taken aback by seeing these optimisations pollute the business logic -- we should try to hide them behind some abstractions or it will become pretty difficult to read Dune's code.

Reply:
(Actually, we already have appendable lists in Stdune, perhaps, we could just switch to them in a few key places.)